File: C++17/백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>
using namespace std;
int arr[15];
int N;

bool Check(int idx) {
	for (int i = 0; i < idx; i++)
		if (arr[i] == arr[idx] || abs(idx - i) == abs(arr[idx] - arr[i]))
			return 0;
	
	return 1;
}

int Queen(int idx) {
	if(N==idx)
		return 1;
	
	int ans = 0;

	for (int i = 0; i < N; i++)
	{
		arr[idx] = i;
		if(!Check(idx)) continue;
		ans += Queen(idx+1);
	}
	return ans;
}

int main() {
	cin >> N;
	cout << Queen(0);
	return 0;
}

File: Java/백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;

public class Main {
    static int[] arr;
    static int s,cnt = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        s = Integer.parseInt(bf.readLine());
        arr = new int[s];
        queen(0);
        System.out.println(cnt);
    }
    public static void queen(int n) {
        if(n == s) {
            cnt++;
            return;
        }
        for(int i = 0; i < s; i++) {
            arr[n] = i;
            if(check(n)) continue;
            queen(n + 1);
        }
    }
    public static boolean check(int n) {
        for (int i = 0; i < n; i++) {
            if (arr[i] == arr[n] || Math.abs(n - i) == Math.abs(arr[n] - arr[i])) return true;
        }
        return false;
    }
}